modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSource.java

```diff
@@ -323,7 +323,6 @@ public void onException(ConnectionException exception) {
     reconnectionAction
         .doOnSuccess(v -> onReconnectionSuccessful())
         .doOnError(this::onReconnectionFailed)
-        .doAfterTerminate(() -> reconnecting.set(false))
```

Comment:
why moving this from here to line 312? also, does the `doOn` vs `doAfter` change really fix the race condition or just makes it less likely?

Reply:
Moving it to line 312 will make it apply only to sources that implement Reconnectable. Other cases are covered by the doWork method (line 548)

Reply:
This case is covered by `org.mule.test.module.extension.source.CustomReconnectionSourceTestCase.successfulCustomReconnection`

Reply:
this seems to be tackling a symptom rather than the root cause. What's the real problem been solved here? Moving this should have no effect, as no two reconnection tasks should be overlapping. If they do, then the bug is a prior race condition

Reply:
What I saw, is that the reconnecting variable is being released (at this point) before the source onStart method ends (which runs asyncrhonously sometimes). And if at the source onStart method a connection exception is notified through the source callback, another reconnection overlaps the current one.

Reply:
right, but the ting is that reconnection is always asynchronous, with the only exception of sources that implement the Reconnectable interface. So maybe it's not about moving this but about how could the overlap happen in the first place. I could well argue that it's really the `doWork` method the one which shouldn't be changing the variable and that the actual bug is the overlap, which you're merely hiding here.

I think you should start by getting good functional tests since the issue seems to revolve exclusively around reconnectable sources which is not something I'm seeing properly replicated on your tests cases.

Reply:
I just pushed the test that I forgot before

Reply:
Change rollbacked, isn't the right fix.